kd_tree.py

```diff
@@ -0,0 +1,149 @@
+import numpy as np
+
+class kdTree:
+    def __init__(self, tree_depth, bounding_box, data, gridId_start):
+        self.root = None
+        self.max_depth = tree_depth
+        self.initial_bb = bounding_box
+        self.data = data
+        self.histogram = None
+        self.bb_collection= []
+        self.gridid_collection = {}
+        self.id_start = gridId_start
+    # =====================================
+    # get a split rule
+    def Get_split(self, level_count, in_bb):
+        # if the counter is 0 or even numbers, we would like to partition a 2-D region along X-axis
+        if level_count == 0 or (level_count & 1) == 0:
+            # get a middle value given min_X and max_X
+            return {'X_Middle_Value' : in_bb[0] + (in_bb[2] - in_bb[0]) / 2 }
+        elif (level_count & 1) == 1:
+            # get a middle value given min_Y and max_Y
+            return { 'Y_Middle_Value' : in_bb[1] + (in_bb[3] - in_bb[1]) / 2 }
+    # =====================================
+    # split the bounding box into two boxes
+    def BB_split(self, in_node, input_BB, level_value):
+        left_up_BB = []
+        right_down_BB = []
+        if (level_value & 1) == 0 or level_value == 0:
+            left_up_BB = [input_BB[0], input_BB[1], input_BB[0] + (input_BB[2] - input_BB[0]) / 2, input_BB[3]]
+            right_down_BB = [input_BB[0] + (input_BB[2] - input_BB[0]) / 2, input_BB[1], input_BB[2], input_BB[3]] 
+        elif (level_value & 1) == 1:
+            left_up_BB = [input_BB[0], input_BB[1] + (input_BB[3] - input_BB[1]) / 2, input_BB[2], input_BB[3]]
+            right_down_BB = [input_BB[0], input_BB[1], input_BB[2], input_BB[1] + (input_BB[3] - input_BB[1]) / 2] 
+        return left_up_BB, right_down_BB
+    # =====================================
+    # build subtrees
+    def build_subtree(self, node, depth_value, input_bb):
+        # split the initial bounding box into two bounding boxes
+        left_up_bb, right_down_bb = self.BB_split(node, input_bb, depth_value)
+        
+        # stop conditions
+        if depth_value >= self.max_depth:
+            node['left'] = left_up_bb 
+            # ==========================================================
+            node['right'] = right_down_bb 
+            return
+        
+        # process left subtrees
+        node['left'] = self.Get_split(depth_value, left_up_bb) # get the split rule
+        self.build_subtree(node['left'], depth_value + 1, left_up_bb)
+        
+        # process right subtrees
+        node['right'] = self.Get_split(depth_value, right_down_bb) # get the split rule
+        self.build_subtree(node['right'], depth_value + 1, right_down_bb)
+        
+    # =====================================
+    # help function
+    def help_fun(self, input_tree, level_val):
```

Comment:
This function name could be more descriptive.

Reply:
re-name the function name.